bridge/github/export.go

```diff
@@ -0,0 +1,731 @@
+func (ge *githubExporter) ExportAll(repo *cache.RepoCache, since time.Time) error {
+	user, err := repo.GetUserIdentity()
+	if err != nil {
+		return err
+	}
+
+	ge.identityToken[user.Id()] = ge.conf[keyToken]
+
+	// get repository node id
+	ge.repositoryID, err = getRepositoryNodeID(
+		ge.conf[keyOwner],
+		ge.conf[keyProject],
+		ge.conf[keyToken],
+	)
+
+	if err != nil {
+		return err
+	}
+
+	allBugsIds := repo.AllBugsIds()
+bugLoop:
+	for _, id := range allBugsIds {
+		b, err := repo.ResolveBug(id)
+		if err != nil {
+			return err
+		}
+
+		snapshot := b.Snapshot()
+
+		// ignore issues created before since date
+		if snapshot.CreatedAt.Before(since) {
+			continue
+		}
+
+		for _, p := range snapshot.Participants {
+			// if we have a token for one of the participants
+			for userId := range ge.identityToken {
+				if p.Id() == userId {
+					// try to export the bug and it associated events
+					if err := ge.exportBug(b, since); err != nil {
+						return err
+					}
+
+					// avoid calling exportBug multiple times for the same bug
+					continue bugLoop
+				}
+			}
+		}
+	}
+
+	fmt.Printf("Successfully exported %d issues and %d labels to Github\n", ge.exportedBugs, ge.exportedLabels)
+	return nil
+}
+
+// exportBug publish bugs and related events
+func (ge *githubExporter) exportBug(b *cache.BugCache, since time.Time) error {
+	snapshot := b.Snapshot()
+
+	var bugGithubID string
+	var bugGithubURL string
+	var bugCreationHash string
+
+	// Special case:
+	// if a user try to export a bug that is not already exported to Github (or imported
+	// from Github) and we do not have the token of the bug author, there is nothing we can do.
+
+	// first operation is always createOp
+	createOp := snapshot.Operations[0].(*bug.CreateOperation)
+	author := createOp.GetAuthor()
+
+	// skip bug if origin is not allowed
+	origin, ok := createOp.GetMetadata(keyOrigin)
+	if ok && !ge.allowOrigin(origin) {
+		// TODO print a warn ?
+		return nil
+	}
+
+	// get github bug ID
+	githubID, ok := createOp.GetMetadata(keyGithubId)
+	if ok {
+		githubURL, ok := createOp.GetMetadata(keyGithubUrl)
+		if !ok {
+			// if we find github ID, github URL must be found too
+			panic("expected to find github issue URL")
+		}
+
+		// will be used to mark operation related to a bug as exported
+		bugGithubID = githubID
+		bugGithubURL = githubURL
+
+	} else {
+		// check that we have a token for operation author
+		client, err := ge.getIdentityClient(author.Id())
+		if err != nil {
+			// if bug is still not exported and we do not have the author stop the execution
+
+			fmt.Println("warning: skipping issue due to missing token for bug creator")
+			// this is not an error, don't export bug
+			return nil
+		}
+
+		// create bug
+		id, url, err := createGithubIssue(client, ge.repositoryID, createOp.Title, createOp.Message)
+		if err != nil {
+			return errors.Wrap(err, "exporting github issue")
+		}
+
+		// incr exported bugs
+		ge.exportedBugs++
```

Comment:
hmmm, if a bug is just updated (no creation, but events added), is it exported ? :thinking:

Reply:
No this is called after a `createGithubIssue` function, maybe we should have a `updatedBugs` variables too ?